dipy/segment/mrf.pyx

```diff
@@ -0,0 +1,657 @@
+#!python
+#cython: boundscheck=False
+#cython: wraparound=False
+#cython: cdivision=True
+import numpy as np
+from dipy.segment.mask import applymask
+from dipy.core.ndindex import ndindex
+from dipy.sims.voxel import add_noise
+cimport cython
+cimport numpy as cnp
+cdef extern from "dpy_math.h" nogil:
+    cdef double NPY_PI
+    cdef double NPY_INFINITY
+    double sqrt(double)
+    double log(double)
+    double exp(double)
+    double fabs(double)
+
+
+class ConstantObservationModel(object):
+    
+    r""" Image segmentation with constant models.

```

Comment:
Since this PR defines the base framework for future tissue segmentation algorithms, I think we should include a detailed description of what each class is supposed to do. For example, for this class, we may say something like:
class ConstantObservationModel(object):
r"""Observation model assuming that the intensity of each class is constant.
The model parameters are the means $\mu_{k}$ and variances $\sigma_{k}$ associated 
with each tissue class.
According to this model, the observed intensity at voxel $x$ is given by
$I(x) = \mu_{k} + \eta_{k}$
where $k$ is the tissue class of voxel $x$, and $\eta_{k}$ is a Gaussian random
variable with zero mean and variance $\sigma_{k}^{2}$.
The observation model is responsible of computing the negative
log-likelihood of observing any given intensity $z$ at each voxel 
$x$ assuming the voxel belongs to each class $k$.
It also provides a default parameter initialization.
"""